File: diff.c
/* Copyright 2019, Mansour Moufid <mansourmoufid@gmail.com> */

#include <assert.h>
#include <stddef.h>

#include "nu.h"
#include "sum.h"

#define co nu_diff_co

static const float nu_diff_co[][5][5] = {
    [3] = {
        { -1.500000f,   2.000000f,  -0.500000f, },
        { -0.500000f,  -0.000000f,   0.500000f, },
        {  0.500000f,  -2.000000f,   1.500000f, },
    },
    [4] = {
        { -1.833333f,   3.000000f,  -1.500000f,   0.333333f, },
        { -0.333333f,  -0.500000f,   1.000000f,  -0.166667f, },
        {  0.166667f,  -1.000000f,   0.500000f,   0.333333f, },
        { -0.333333f,   1.500000f,  -3.000000f,   1.833333f, },
    },
    [5] = {
        { -2.083333f,   4.000000f,  -3.000000f,   1.333333f,  -0.250000f, },
        { -0.250000f,  -0.833333f,   1.500000f,  -0.500000f,   0.083333f, },
        {  0.083333f,  -0.666667f,   0.000000f,   0.666667f,  -0.083333f, },
        { -0.083333f,   0.500000f,  -1.500000f,   0.833333f,   0.250000f, },
        {  0.250000f,  -1.333333f,   3.000000f,  -4.000000f,   2.083333f, },
    },
};

const size_t nu_diff_kmax = 5;

void
nu_diff(float h, size_t k, float z[], const float x[], size_t n)
{
    assert(h > 0.f);
    assert(k > 1 && k <= nu_diff_kmax);
    float w = 1.f / h;
    for (size_t i = 0; i < k / 2; i++) {
        size_t j = 0;
        assert(j <= n - k);
        z[i] = w * _nu_dot(co[k][i], &(x[j]), k);
    }
    for (size_t i = k / 2; i < n - k / 2; i++) {
        size_t j = i - k / 2;
        assert(j <= n - k);
        z[i] = w * _nu_dot(co[k][k / 2], &(x[j]), k);
    }
    for (size_t i = n - k / 2; i < n; i++) {
        size_t j = n - k;
        assert(j <= n - k);
        z[i] = w * _nu_dot(co[k][k - (n - i)], &(x[j]), k);
    }
}

File: diff.py
# Copyright 2019, 2020, Mansour Moufid <mansourmoufid@gmail.com>

from __future__ import print_function

try:
    from StringIO import StringIO
except:
    from io import StringIO
import string
import sys


def ks(n):
    return map(float, range(-((n + 1) // 2) + 1, n // 2 + 1))


cos = {
    3: [[x - 0.5, -2 * x, x + 0.5] for x in ks(3)],
    4: [[-(3 * x**2 - 6 * x + 2) / 6,
          (3 * x**2 - 4 * x - 1) / 2,
         -(3 * x**2 - 2 * x - 2) / 2,
          (3 * x**2 - 1) / 6] for x in ks(4)],
    5: [[ (2 * x**3 - 3 * x**2 - x + 1) / 12,
         -(4 * x**3 - 3 * x**2 - 8 * x + 4) / 6,
          (2 * x**3 - 5 * x) / 2,
         -(4 * x**3 + 3 * x**2 - 8 * x - 4) / 6,
          (2 * x**3 + 3 * x**2 - x - 1) / 12] for x in ks(5)],
}

indent = ' ' * 4

co = StringIO()
nco = max(cos.keys())

print('{', file=co)
for k in cos:
    print('{}[{}] = {{'.format(indent, k), file=co)
    for x in cos[k]:
        co.write(2 * indent + '{')
        for f in x:
            co.write('{:10.6f}f, '.format(f))
        print('},', file=co)
    print(indent + '},', file=co)
co.write('}')

for line in sys.stdin:
    template = string.Template(line)
    out = template.substitute({
        'co': co.getvalue(),
        'nco': nco,
    })
    sys.stdout.write(out)
